test/test-lex.cpp

```diff
@@ -649,10 +651,77 @@ TEST_F(test_lex, lex_strings) {
                                 offsets_matcher(input, 0, 14))));
       });
 
-  // TODO(strager): Report invalid hex escape sequences. For example:
-  //
-  // "hello\x1qworld"
-  // '\x'
+  this->check_tokens_with_errors(
+      u8"'\\x", {token_type::string},
+      [](padded_string_view input, const auto& errors) {
+        EXPECT_THAT(errors,
+                    ElementsAre(
+                      ERROR_TYPE_FIELD(
+                        error_invalid_hex_escape_sequence, escape_sequence,
+                        offsets_matcher(input, 1, 4)),
+                      ERROR_TYPE_FIELD(
+                        error_unclosed_string_literal, string_literal,
+                        offsets_matcher(input, 0, 3))
+                    )
+        );
+      });
+
+  this->check_tokens_with_errors(
+      u8"'\\x1", {token_type::string},
+      [](padded_string_view input, const auto& errors) {
+        EXPECT_THAT(errors,
+                    ElementsAre(
+                      ERROR_TYPE_FIELD(
+                        error_invalid_hex_escape_sequence, escape_sequence,
+                        offsets_matcher(input, 1, 5)),
+                      ERROR_TYPE_FIELD(
+                        error_unclosed_string_literal, string_literal,
+                        offsets_matcher(input, 0, 4))
+                    )
+        );
+      });
+
+  this->check_tokens_with_errors(
+      u8"'\\x'", {token_type::string},
+      [](padded_string_view input, const auto& errors) {
+        EXPECT_THAT(errors,
+                    ElementsAre(
+                      ERROR_TYPE_FIELD(
+                        error_invalid_hex_escape_sequence, escape_sequence,
+                        offsets_matcher(input, 1, 4))
+                    )
+        );
+      });
+
+  this->check_tokens_with_errors(
+      u8"'\\x\\xyz'", {token_type::string},
+      [](padded_string_view input, const auto& errors) {
+        EXPECT_THAT(errors,
+                    ElementsAre(
+                      ERROR_TYPE_FIELD(
+                        error_invalid_hex_escape_sequence, escape_sequence,
+                        offsets_matcher(input, 1, 4)),
+                      ERROR_TYPE_FIELD(
+                        error_invalid_hex_escape_sequence, escape_sequence,
+                        offsets_matcher(input, 3, 6))
```

Comment:
Hmm, these errors overlap. Is that a good idea? Not sure.

Reply:
Some overlapping may be tough to avoid. For example, should invalid hex escape sequence also be raised inside an unclosed string?

Anyway, this error should never overlap anymore since it's been changed to always output `\x`.